### PR TITLE
Bug 1875770: Test for PVC clone

### PR DESCRIPTION
**Pull request by cloudbehl**
Dependency: https://github.com/openshift/console/pull/6436
Signed-off-by: Ankush Behl <cloudbehl@gmail.com>

### CODE REVIEW

File: frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      .then((res) => {
+        const pvc = JSON.parse(res);
+        cy.get(PVCDetails.name).contains(pvc.metadata.name);
+        cy.get(PVCDetails.namespace).contains(pvc.metadata.namespace);
+        cy.byTestID('pvc-requested-capacity').contains(`${cloneSize} GiB`);
+      });
+  });
+
+  it('Lists Clone', () => {
+    cy.clickNavLink((['Persistent Volume Claims'] as unknown) as [string, string]);
```
Comment (bipuladh) [inline frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts:56]:
```suggestion
    cy.clickNavLink((['Persistent Volume Claims']);
```
This issue has been fixed already.

File: frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    modal.submit();
+    modal.shouldBeClosed();
+    cy.location('pathname').should('include', `persistentvolumeclaims/${PVC.metadata.name}-clone`);
+    detailsPage.titleShouldContain(`${PVC.metadata.name}-clone`);
+    cy.exec(`oc get pvc ${PVC.metadata.name}-clone -n ${testName} -o json`)
+      .its('stdout')
+      .then((res) => {
+        const pvc = JSON.parse(res);
+        cy.get(PVCDetails.name).contains(pvc.metadata.name);
+        cy.get(PVCDetails.namespace).contains(pvc.metadata.namespace);
```
Comment (bipuladh) [inline frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts:50]: Looks like we are using the same selector for snapshot too, maybe we can refactor it to a common selector `DetailsPage` perhaps...?

**Review by bipuladh — CHANGES_REQUESTED** [review]

**Comment by cloudbehl** [comment]
@dtaylor113 can you please review?

File: frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    modal.shouldBeOpened();
+    modal.submitShouldBeEnabled();
+    cy.byTestID('input-request-size')
+      .clear()
+      .type(cloneSize);
+    modal.submit();
+    modal.shouldBeClosed();
+    cy.location('pathname').should('include', `persistentvolumeclaims/${PVC.metadata.name}-clone`);
+    detailsPage.titleShouldContain(`${PVC.metadata.name}-clone`);
+    cy.exec(`oc get pvc ${PVC.metadata.name}-clone -n ${testName} -o json`)
```
Comment (bipuladh) [inline frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts:45]: If you get an exit code as 0, then the request should be successful.

**Review by bipuladh — CHANGES_REQUESTED** [review]

**Review by bipuladh — APPROVED** [review]

File: frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      .its('stdout')
+      .then((res) => {
+        const pvc = JSON.parse(res);
+        cy.get(DetailsPageSelector.name).contains(pvc.metadata.name);
+        cy.get(DetailsPageSelector.namespace).contains(pvc.metadata.namespace);
+        cy.byTestID('pvc-requested-capacity').contains(`${cloneSize} GiB`);
+      });
```
Comment (bipuladh) [inline frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts:52]: I don't think this call is required as most of these values can be deciphered from the PVC request object, `cloneSize` and `testName`.

**Review by bipuladh — APPROVED** [review]

File: frontend/public/components/utils/request-size-input.tsx [hunk 5]
```diff
@@ -67,4 +68,5 @@ export type RequestSizeInputProps = {
   minValue?: number;
   inputClassName?: string;
   inputID?: string;
+  dataTestId?: string;
```
Comment (spadgett) [inline frontend/public/components/utils/request-size-input.tsx:71]:
```suggestion
  testID?: string;
```

File: frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+const cloneName = `${PVC.metadata.name}-clone`;
+const cloneSize = '2';
+
+describe('Clone Tests', () => {
+  before(() => {
+    cy.login();
+    cy.createProject(testName);
+    cy.exec(`echo '${JSON.stringify(PVC)}' | oc apply -n ${testName} -f -`);
+    cy.exec(`echo '${JSON.stringify(testerDeployment)}' | oc apply -n ${testName} -f -`);
+    cy.clickNavLink(['Storage', 'Persistent Volume Claims']);
```
Comment (dtaylor113) [inline frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts:18]: Just a heads up that when #6521 merges, this should be changed to `nav.sidenav.clickNavLink(..`.   A bunch of new helper methods added with `nav.sidenav`.

File: frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+      .then((res) => {
+        const pvc = JSON.parse(res);
+        cy.get(DetailsPageSelector.name).contains(pvc.metadata.name);
+        cy.get(DetailsPageSelector.namespace).contains(pvc.metadata.namespace);
+        cy.byTestID('pvc-requested-capacity').contains(`${cloneSize} GiB`);
+      });
+  });
+
+  it('Lists Clone', () => {
+    cy.clickNavLink(['Persistent Volume Claims']);
```
Comment (dtaylor113) [inline frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts:56]: Just a heads up that when #6521 merges, this should be changed to `nav.sidenav.clickNavLink(..`

**Comment by dtaylor113** [comment]
Looks like CI timed out waiting for 'Pending' to become 'Bound':

image

**Review by dtaylor113 — CHANGES_REQUESTED** [review]
Looks like CI timed out waiting for 'Pending' to become 'Bound'.   Do we need to increase timeout?

File: frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts (hunk 7 above)
Comment (cloudbehl) [reply frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts]: Ack!

File: frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts (hunk 6 above)
Comment (cloudbehl) [reply frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts]: ack!

**Comment by cloudbehl** [comment]
@dtaylor113 can you review it once again?

**Review by dtaylor113 — APPROVED** [review]
lgtm, thanks

File: frontend/public/components/utils/request-size-input.tsx [hunk 8]
```diff
@@ -67,4 +68,5 @@ export type RequestSizeInputProps = {
   minValue?: number;
   inputClassName?: string;
   inputID?: string;
+  testId?: string;
```
Comment (spadgett) [inline frontend/public/components/utils/request-size-input.tsx:71]:
```suggestion
  testID?: string;
```

Comment (cloudbehl) [reply frontend/public/components/utils/request-size-input.tsx]: ack!

File: frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts (hunk 7 above)
Comment (cloudbehl) [reply frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts]: Done

File: frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts (hunk 6 above)
Comment (cloudbehl) [reply frontend/packages/integration-tests-cypress/tests/storage/clone.spec.ts]: Done

**Comment by cloudbehl** [comment]
@dtaylor113  Updated the PR to address sam's comment & also added addressed https://github.com/openshift/console/pull/6528#pullrequestreview-486949555

**Review by spadgett — APPROVED** [review]
/approve

**Comment by spadgett** [comment]
Adding lgtm on behalf of @dtaylor113
/lgtm

@dtaylor113 note nothing can come after lgtm on the line with the bot command :)

**Pull request merged**